### PR TITLE
723 change local setup for better devving

### PR DESCRIPTION
**Pull request by RobertM905**
### Context

We are currently using docker in our CI/CD pipeline, but we have not leveraged docker and docker-compose to help during local development. This PR hopes to address this.

### Changes proposed in this pull request

This PR proposes the addition of a Makefile to the project that contains a number of helpful commands for local development including: 
- `make test` : builds a fresh docker container to run the test suite.
- `make shell` : builds a fresh docker container and starts an interactive shell within the container.
- `make serve` builds a fresh docker container that serves the rails application to localhost for development.

This PR also edits the current `database.yml` config file to remove database defaults that should be passed in via environment variables. 
This change requires anyone running the application to have to specify env variables for: 
`DB_HOSTNAME`
`DB_PORT`
`DB_DATABASE`

Unless using the `Makefile` or `docker-compose` commands.

### Guidance to review

 - Please review the Makefile addition. 
 - Review the changes to the `database.yml` file.
 - Review the changes to the `docker-compose.yml` file.

### Link to Trello card

[723 - Change local setup so devs can easily build test and deploy the service](https://trello.com/c/8dfnU6LH/723-change-local-setup-so-devs-can-easily-build-test-and-deploy-the-service)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,21 @@
+.PHONY: test
+
+test:
+	docker-compose down
+	docker-compose build
+	docker-compose run --rm web bundle exec rails db:test:prepare
+	docker-compose run --rm web bundle exec rspec
```
Comment (duncanjbrown) [inline Makefile]: Should we run `bundle exec rake` here rather than just `rspec`, as it'll include Rubocop and ERB lint?

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,21 @@
+.PHONY: test
+
+test:
+	docker-compose down
```
Comment (duncanjbrown) [inline Makefile]: Is there any reason to run `down` here? `run` will create a new container, and it's possible to `build` while containers are running

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+	docker-compose build
+	docker-compose run --rm web bundle exec rails db:test:prepare
+	docker-compose run --rm web bundle exec rspec
+
+.PHONY: shell
+
+shell:
+	docker-compose down
```
Comment (duncanjbrown) [inline Makefile]: Is `down` needed here either?

**Comment by kjdchapman** [comment]
> Ought we to start mounting ".:/app" so that we can dev in Docker?

This is a good idea, I'd happily add this to this PR.

File: Makefile (hunk 1 above)
Comment (kjdchapman) [reply Makefile]: Generally I've linted as part of a separate command (`make lint` or `make check`), as I find myself running the tests more often than the linting.

Maybe the linting is so quick I shouldn't bother separate the two.

I may have had some trouble running rake in this context, but I can't remember what it was - will look again, or @RobertM905 can have a go?

File: Makefile (hunk 2 above)
Comment (kjdchapman) [reply Makefile]: I think I introduced this whilst I was still playing with the `postgres` image, and it's not necessary.

**Comment by duncanjbrown** [comment]
Following team discussions this morning, I've

- added make targets for CI, which I've namespaced — comments/opinions welcome — under `ci.` in the `Makefile`
- updated the pipeline to use the new targets (see the "Set `COMPOSE_FILE` in pipeline" commit for some extra context)
- added a default `make help` target which prints helpful help:
Screenshot 2019-08-13 at 12 42 45

- hopefully clarified the docker database situation (see commits)

@tpinney, are you happy with the change in the pipeline?

File: config/database.yml [hunk 4]
```diff
@@ -1,25 +1,16 @@
...
   username: <%= ENV['DB_USERNAME'] %>
   password: <%= ENV['DB_PASSWORD'] %>
-  host: <%= ENV['DB_HOSTNAME'] || 'localhost' %>
-  port: <%= ENV['DB_PORT'] || '5432' %>
+  host: <%= ENV['DB_HOSTNAME'] %>
+  port: <%= ENV['DB_PORT'] %>
 
 development:
   <<: *default
-  database: <%= ENV['DB_DATABASE'] || 'bat_apply_dev' %>
```
Comment (tpinney) [inline config/database.yml]: The deployment pipeline requires the DB_DATABASE, so removing this, as proposed, will break stages further down the pipeline. What is the reasoning behind removing the override functionality?

File: config/database.yml [hunk 5]
```diff
@@ -1,25 +1,16 @@
...
+  database: 'bat_apply_dev'
 
-# Warning: The database defined as "test" will be erased and
-# re-generated from your development database when you run "rake".
-# Do not set this db to the same as development or production.
 test:
   <<: *default
-  database: <%= ENV['DB_DATABASE'] || 'bat_apply_test' %>
```
Comment (tpinney) [inline config/database.yml]: Same query as for "development" above.

**Comment by duncanjbrown** [comment]
Just discussed with @tpinney. We need to be able to override the DB_DATABASE when deploying. At the moment, though, we deploy the app in development mode, which means the current pipeline depends on the ENV var being referenced in `database.yml`. We _should_ be deploying in production mode, though, so I'm going to try:

- changing the pipeline to deploy in prod mode
- adding a key to the `database.yml` file to configure the prod database name via the DB_DATABASE env variable

**Comment by duncanjbrown** [comment]
I've taken a quick look, and changing the pipeline to run in production is a significant piece of work. I've 

- restored the `DB_DATABASE` override in `development` for now, but left it hardcoded in `test`
- created a card for it https://trello.com/c/YovhLqlr/772-run-the-app-in-production-mode-on-azure

**Comment by duncanjbrown** [comment]
This should be ready for a final review

**Pull request merged**